Title: AiModel.stream() missing middleware injection and closure scoping bugs in provider chatStream/chat methods

What are the steps to reproduce this issue?
Configure an AiModel with middleware (e.g. LoggingMiddleware) and call model.stream() with a callback.
Use DockerModelRunner provider for streaming — observe console errors during stream chunk parsing.
Use any OpenAI-compatible provider with tool calling in both chat() and chatStream() — middleware may fail to fire on tool call iterations.
What happens?
AiModel.stream() does not inject middleware into the chatRequest before calling chatStream(). Compare with run() which injects both agent and model middleware (lines 148-154). The streaming path is completely unaware of middleware — beforeLLMCall, afterLLMCall, and all tool call middleware hooks never fire.

DockerModelRunnerService.chat() and chatStream() pass callback and chatRequest unscoped into a closure passed to retryOnModelLoading(). BoxLang closures have their own arguments scope, so the unscoped references resolve against the wrong scope, causing: Error getting method [callback] for class [ortus.boxlang.runtime.scopes.ArgumentsScope]

OpenAIService.chatStream() passes callback and chatRequest unscoped to sendStreamRequest() on line 399, and uses unscoped chatRequest inside the .each() tool-calling closure (lines 452-453). Same ArgumentsScope resolution failure.

OpenAIService.chat() uses unscoped chatRequest inside nested .each() closures for tool calling (lines 257, 262).

CohereService.chat() uses unscoped chatRequest inside the .map() tool-calling closure (line 282).

What were you expecting to happen?
AiModel.stream() should inject middleware identically to run() so that beforeLLMCall, afterLLMCall, beforeToolCall, afterToolCall, and wrapToolCall/wrapLLMCall hooks all fire during streaming.
All closure references to arguments.chatRequest and arguments.callback should be captured into local variables before entering closures to avoid BoxLang scope chain resolution failures.
Any logs, error output, etc?
[ERROR] [AI] Error parsing stream chunk: [Error getting method [callback] for class [ortus.boxlang.runtime.scopes.ArgumentsScope]]

Repeats for every stream chunk when using DockerModelRunner.

Any other comments?
Files affected:

src/main/bx/models/runnables/AiModel.bx — stream() missing middleware injection
src/main/bx/models/providers/DockerModelRunnerService.bx — unscoped args in retryOnModelLoading closure
src/main/bx/models/providers/OpenAIService.bx — unscoped args in chatStream() and chat() tool closures
src/main/bx/models/providers/CohereService.bx — unscoped args in chat() tool closure

What versions are you using?
Operating System: macOS (Darwin 25.3.0)
Package Version: bx-ai 3.0

--- a/src/main/bx/models/providers/CohereService.bx
+++ b/src/main/bx/models/providers/CohereService.bx
@@ -270,12 +270,13 @@ class extends="OpenAIService" implements="IAiChatService, IAiEmbeddingsService" 
 			}
 
 			// Execute each tool call
+			var thisChatRequest = arguments.chatRequest;
 			var toolResults = result.tool_calls.map( toolCall => {
 				// Find the tool in our original tools list
 				var tool = tools.filter( t => t.getSchema().function.name == toolCall.name ).first();
 
 				// Invoke the tool with the parameters
-				var toolOutput = tool.invoke( toolCall.parameters, chatRequest );
+				var toolOutput = tool.invoke( toolCall.parameters, thisChatRequest );
 
 				return {
 					"call": toolCall,

--- a/src/main/bx/models/providers/DockerModelRunnerService.bx
+++ b/src/main/bx/models/providers/DockerModelRunnerService.bx
@@ -115,8 +115,11 @@ class extends = "OpenAICompatibleService" implements="IAiChatService, IAiEmbeddi
 	){
 		// Seed the chat request with defaults
 		arguments.chatRequest.mergeServiceParams( static.DEFAULT_CHAT_PARAMS )
+		// Capture arguments in local scope so the closure doesn't lose them
+		var thisChatRequest      = arguments.chatRequest
+		var thisInteractionCount = arguments.interactionCount
 		return retryOnModelLoading( () => {
-			return super.chat( chatRequest = chatRequest, interactionCount = interactionCount )
+			return super.chat( chatRequest = thisChatRequest, interactionCount = thisInteractionCount )
 		} )
 	}
 
@@ -136,8 +139,12 @@ class extends = "OpenAICompatibleService" implements="IAiChatService, IAiEmbeddi
 	){
 		// Seed the chat request with defaults
 		arguments.chatRequest.mergeServiceParams( static.DEFAULT_CHAT_PARAMS )
+		// Capture arguments in local scope so the closure doesn't lose them
+		var thisChatRequest      = arguments.chatRequest
+		var thisCallback         = arguments.callback
+		var thisInteractionCount = arguments.interactionCount
 		return retryOnModelLoading( () => {
-			return super.chatStream( chatRequest = chatRequest, callback = callback, interactionCount = interactionCount )
+			return super.chatStream( chatRequest = thisChatRequest, callback = thisCallback, interactionCount = thisInteractionCount )
 		} )
 	}
 

--- a/src/main/bx/models/providers/OpenAIService.bx
+++ b/src/main/bx/models/providers/OpenAIService.bx
@@ -249,14 +249,15 @@ class extends="BaseService" implements="IAiChatService, IAiEmbeddingsService" {
 		 */
 
 		// Capture chatRequest in local scope for use in nested closures
+		var thisChatRequest = arguments.chatRequest
 		var newMessages = arguments.chatRequest.getMessages().map( message -> message )
 		result.choices.each( ( choice, index ) => {
-			var targetChatRequest = chatRequest
+			var targetChatRequest = thisChatRequest
 			// add the tool call into our message history
 			newMessages.append( choice.message )
 			// find the tool, invoke it, append the result to the chat history
 			choice.message.tool_calls.each( ( toolCall, i ) => {
-				var toolAttempt = chatRequest.getTool( toolCall.function.name )
+				var toolAttempt = thisChatRequest.getTool( toolCall.function.name )
 				if( toolAttempt.isPresent() ){
 					var tool = toolAttempt.get();
 
@@ -393,7 +394,7 @@ class extends="BaseService" implements="IAiChatService, IAiEmbeddingsService" {
 		// -------------------------------------------------------------------------------------------
 		// --- Stream the chat response, with streaming tool call support ---
 		// -------------------------------------------------------------------------------------------
-		var streamState = sendStreamRequest( chatRequest, dataPacket, callback )
+		var streamState = sendStreamRequest( arguments.chatRequest, dataPacket, arguments.callback )
 
 		// -------------------------------------------------------------------------------------------
 		// --- Fire local callback for post mods ---
@@ -445,9 +446,11 @@ class extends="BaseService" implements="IAiChatService, IAiEmbeddingsService" {
 		assistantMessage.tool_calls = streamState.toolCalls
 		newMessages.append( assistantMessage )
 
+		var thisChatRequest = arguments.chatRequest
+		var thisCallback    = arguments.callback
 		streamState.toolCalls.each( toolCall => {
-			var targetChatRequest = chatRequest
-			var toolAttempt       = chatRequest.getTool( toolCall.function.name ?: "" )
+			var targetChatRequest = thisChatRequest
+			var toolAttempt       = thisChatRequest.getTool( toolCall.function.name ?: "" )
 			var toolArguments     = {}
 
 			try {
@@ -533,8 +536,8 @@ class extends="BaseService" implements="IAiChatService, IAiEmbeddingsService" {
 		} );
 
 		// Continue streaming the final answer after tool execution
-		arguments.chatRequest.setMessages( newMessages );
-		chatStream( arguments.chatRequest, arguments.callback, arguments.interactionCount + 1 );
+		thisChatRequest.setMessages( newMessages );
+		chatStream( thisChatRequest, thisCallback, arguments.interactionCount + 1 );
 	}
 
 	/**

--- a/src/main/bx/models/runnables/AiModel.bx
+++ b/src/main/bx/models/runnables/AiModel.bx
@@ -232,6 +232,19 @@ class extends="AiBaseRunnable" {
 		)
 
 		// -------------------------------------------------------------------------------------------
+		// --- Prepare middleware chain ---
+		// Agent middleware (if this model is being run by an agent) is added at the AiAgent level,
+		// so it wraps the entire model execution including all middleware and the service call.
+		// -------------------------------------------------------------------------------------------
+		// Inject middleware: agent middleware first (from options), then model's own middleware
+		if ( effectiveOptions.keyExists( "_agentMiddleware" ) && isArray( effectiveOptions._agentMiddleware ) ) {
+			chatRequest.addMiddleware( effectiveOptions._agentMiddleware )
+		}
+		if ( !variables.middleware.isEmpty() ) {
+			chatRequest.addMiddleware( variables.middleware )
+		}
+
+		// -------------------------------------------------------------------------------------------
 		// --- Announce before invocation (global event) ---
 		// -------------------------------------------------------------------------------------------
         BoxAnnounce( "beforeAIModelInvoke", { model: this, chatRequest: chatRequest } )

